jscomp/bsb/bsb_helper_main.ml

```diff
@@ -17,24 +17,101 @@
+
+let includes :  _ list ref = ref []
+
+let add_include cwd dir =
+  includes := process_include cwd dir :: !includes
+
+let batch_files = ref []
+let collect_file name =
+  batch_files := name :: !batch_files
 
 let output_prefix = ref None
 let dev_group = ref 0
-let annoymous other = ()
+let link_flag = ref None
+let pack_flag = ref None
+
+let anonymous filename =
+  collect_file filename
 let usage = "Usage: bsb_helper.exe [options] \nOptions are:"
 let () =
     Arg.parse [
     "-oprefix", Arg.String (fun x -> output_prefix := Some x),
     " Set output prefix for -bs-MD (internal use)";
-    "-g", Arg.Int (fun i -> dev_group := i ), 
+    "-g", Arg.Int (fun i -> dev_group := i ),
     " Set the dev group (default to be 0)"
     ;
-    "-MD", Arg.String (fun x -> Depends_post_process.handle_bin_depfile !output_prefix x !dev_group ),
-    " (internnal)Generate dep file for ninja format(from .ml[i]deps)";
+    "-MD", Arg.String (fun x -> Depends_post_process.handle_bin_depfile ~compilation_kind:Js !output_prefix x !dev_group ),
+    " (internal)Generate dep file for ninja format(from .ml[i]deps)";
+    "-MD-bytecode", Arg.String (fun x -> Depends_post_process.handle_bin_depfile ~compilation_kind:Bytecode !output_prefix x !dev_group ),
+    " (internal)Generate dep file for ninja format(from .ml[i]deps)";
+    "-MD-native", Arg.String (fun x -> Depends_post_process.handle_bin_depfile ~compilation_kind:Native !output_prefix x !dev_group ),
+    " (internal)Generate dep file for ninja format(from .ml[i]deps)";
+
+    (* Used for linking *)
+    "-bs-main", (Arg.String set_main_module),
+    " set the main entry module. Only used in conjunction with -link-bytecode and -link-native";
+
+    "-I",  (Arg.String (add_include (Sys.getcwd ()))),
+    " add dir to search path for the linker and packer";
+
+    "-link-bytecode", (Arg.String (fun x -> link_flag := Some (Bsb_helper_linker.LinkBytecode x))),
+    " link bytecode files into an executable";
+
+    "-link-native", (Arg.String (fun x -> link_flag := Some (Bsb_helper_linker.LinkNative x))),
+    " link native files into an executable";
+
+    "-pack-native-library", (Arg.Unit (fun () -> pack_flag := Some Bsb_helper_packer.PackNative)),
+    " pack native files (cmx) into a library file (cmxa)";
+
+    "-pack-bytecode-library", (Arg.Unit (fun () -> pack_flag := Some Bsb_helper_packer.PackBytecode)),
+    " pack bytecode files (cmo) into a library file (cma)";
+    ] anonymous usage
+
+(**
+ * This makes bsb_helper act as an ocaml linker where we automatically figure
+ * out the dependencies graph to do a topological sort before calling ocamlc.
+ *
+ * Below we check link_flag first, then check pack_flag.
+ * *)
+let () = match !link_flag with
+| None -> ()
+| Some link_byte_or_native -> 
+  begin match !main_module with
+  | None -> failwith "linking needs a main module"
+  | Some main_module ->
+    Bsb_helper_linker.link 
+      link_byte_or_native
+      ~main_module:main_module
```

Comment:
I'm not sure what you mean... This only happens if `link_flag` contained `Some _` right? So only if the user supplied the parameter.